src/components/validation/runner/telemetry.rs

```diff
@@ -57,31 +72,80 @@ impl Telemetry {
 
             let mut telemetry_events = Vec::new();
-            loop {
+            'outer: loop {
                 select! {
-                    _ = telemetry_shutdown_handle.wait() => break,
+                    _ = telemetry_shutdown_handle.wait() => {
+                        // After we receive the shutdown signal, we need to wait
+                        // for two event emissions from the internal_metrics
+                        // source. This is to ensure that we've received all the
+                        // events from the components that we're testing.
+                        //
+                        // We need exactly two because the internal_metrics
+                        // source does not emit component events until after the
+                        // component_received_events_total metric has been
+                        // emitted. Thus, two events ensure that all component
+                        // events have been emitted.
+
+                        debug!("Telemetry: waiting for final internal_metrics events before shutting down.");
+
+                        let mut events_seen = 0;
+                        let current_time = chrono::Utc::now();
+
+                        loop {
+                            match &rx.recv().await {
+                                None => break 'outer,
+                                Some(telemetry_event) => {
+                                    telemetry_events.push(telemetry_event.clone());
+                                    if let Event::Metric(metric) = telemetry_event {
+                                        if let Some(tags) = metric.tags() {
+                                            if metric.name() == INTERNAL_METRICS_SHUTDOWN_EVENT &&
+                                            tags.get("component_name") == Some(INTERNAL_LOGS_KEY) &&
+                                            metric.data().timestamp().unwrap() > &current_time {
+                                                debug!("Telemetry: processed one component_received_events_total event.");
+
+                                                events_seen += 1;
+                                                if events_seen == 2 {
+                                                    break 'outer;
+                                                }
+                                            }
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    },
                     maybe_telemetry_event = rx.recv() => match maybe_telemetry_event {
                         None => break,
                         Some(telemetry_event) => telemetry_events.push(telemetry_event),
                     },
                 }
             }
 
+            grpc_task_coordinator.shutdown().await;
+            debug!("GRPC task(s) have been shutdown.");
```

Comment:
Non-blocking nit: should be `gRPC`.

Reply:
Yep--I got a complaint from one of the linters for not capitalizing the first letter of the message.

Reply:
Wahhhh, silly linter... whatever. :P

Reply:
Not the spell checker right? Probably the one for events?